Make TensorImpl custom impls call their defaults

bdhirsh opened the pull request:
It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81302
* #81301
* #81238
* __->__ #81236
* #81300
* #82453

bdhirsh committed:
Make TensorImpl custom impls call their defaults

[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"




[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"




[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"

It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.




[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"

It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.




[ghstack-poisoned]

ezyang reviewed (approved):
thanks, I think functorch, xla and torchdistx are the most likely out of tree stuff to also check

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"

It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.




[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"

It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.




[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"

It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.




[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"

It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.




[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"

It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.




[ghstack-poisoned]

bdhirsh committed:
Update on "Make TensorImpl custom impls call their defaults"

It looks like I actually need this for functionalization: it's possible to have a `TensorImpl` with a custom sizes policy, that:
- does not have the `Python` dispatch key on it
- is not a custom `TensorImpl` subclass

So this PR lets the custom implementations on `TensorImpl` fall through to the default, and leaves it up to the subclasses of `TensorImpl` to error appropriately.

I added the checks directly to a few subclasses, but waiting on CI to reveal any issues.




[ghstack-poisoned]
